### PR TITLE
Canvas: Add support to rotate a group of elements

### LINKED ISSUE #84961: Canvas: Consider adding support for rotating a group of elements
Currently when selecting a group of elements the rotatable control appears but does not actually apply any rotation to the selected elements.

Look into supporting this functionality. The movie below show cases what this might look like using the `rotateGroup` event

https://github.com/grafana/grafana/assets/22381771/ad3b99e5-8a12-47d1-bb5d-b21d451a5279

### PR DESCRIPTION
**Pull request by nmarrs**
Add support to rotate elements in a group - current UX indicates this is possible so seems like a bug / not a great experience for users

Before

https://github.com/grafana/grafana/assets/22381771/81571ac7-e15f-4672-802a-d10325adfb4e

After

https://github.com/grafana/grafana/assets/22381771/b42feabf-58c1-462f-a4ec-d192ec15d3fc

After (with connections!) - same behavior as rotating a single element / no regressions

https://github.com/grafana/grafana/assets/22381771/71942ccd-e3a7-46ef-ad94-7e28dcf74eca

Fixes https://github.com/grafana/grafana/issues/84961

### CODE REVIEW

File: public/app/features/canvas/runtime/element.tsx [hunk 1]
```diff
@@ -489,11 +489,14 @@ export class ElementState implements LayerElement {
   };
 
   applyRotate = (event: OnRotate) => {
-    const absoluteRotationDegree = event.absoluteRotation;
-
+    const rotationDelta = event.delta;
     const placement = this.options.placement!;
+    const placementRotation = placement.rotation ?? 0;
+
+    const calculatedRotation = placementRotation + rotationDelta;
```
Comment (nmarrs) [inline public/app/features/canvas/runtime/element.tsx:496]: This change impacts both single element and group element rotation; TLDR by applying the relative delta we can have a solution that works for both single element + group of element rotation (see commit history for additional context if interested)

**Comment by drew08t** [comment]
The rotation works as expected for me, but I'm seeing some strange things happening to styling on multiple panels. Notice the background colors:
image
Not sure if it's a regression from this PR (doubt it) or something else is happening.

Also, a couple of notes:
1. In general this is a good step forward, but I think we need to revisit group rotation in the near future. For example, should we have all elements rotate about a single point? Obviously the handling of anchors needs to be improved as well in a follow-up.
2. Why are there two rotation anchors?

**Comment by nmarrs** [comment]
> The rotation works as expected for me, but I'm seeing some strange things happening to styling on multiple panels. Notice the background colors: image Not sure if it's a regression from this PR (doubt it) or something else is happening.

Hm I noticed regression as well, I think it is pretty recent regression that we probably will want to get to the bottom of sooner than later

> 1. In general this is a good step forward, but I think we need to revisit group rotation in the near future. For example, should we have all elements rotate about a single point? Obviously the handling of anchors needs to be improved as well in a follow-up.

How else might a group of elements rotate? 🤔 I think moveable has support for changing the relative "center" of rotation but not sure what use cases / value this would add in this case

Re: connection anchor rendering issue we have [this issue to follow up on this](https://github.com/grafana/grafana/issues/84957)

> 2. Why are there two rotation control handles?

See [this issue](https://github.com/grafana/grafana/issues/84959) and [the PR fix](https://github.com/grafana/grafana/pull/85690) that addressed it :)

**Comment by drew08t** [comment]
> > The rotation works as expected for me, but I'm seeing some strange things happening to styling on multiple panels. Notice the background colors: image Not sure if it's a regression from this PR (doubt it) or something else is happening.
> 
> Hm I noticed regression as well, I think it is pretty recent regression that we probably will want to get to the bottom of sooner than later
> 
> > 1. In general this is a good step forward, but I think we need to revisit group rotation in the near future. For example, should we have all elements rotate about a single point? Obviously the handling of anchors needs to be improved as well in a follow-up.
> 
> How else might a group of elements rotate? 🤔 I think moveable has support for changing the relative "center" of rotation but not sure what use cases / value this would add in this case
> 
> Re: connection anchor rendering issue we have [this issue to follow up on this](https://github.com/grafana/grafana/issues/84957)
> 
> > 2. Why are there two rotation control handles?
> 
> See [this issue](https://github.com/grafana/grafana/issues/84959) and [the PR fix](https://github.com/grafana/grafana/pull/85690) that addressed it :)

For me, it's more intuitive to rotate a group of elements about a common point, perhaps the centroid of the selected bounding boxes could work.

**Review by drew08t — APPROVED** [review]
This is definitely a great step in the right direction! Nice work!

I created a [follow-up issue to fix the element background colors](https://github.com/grafana/grafana/issues/87359).

I tested main and the regression is present there as well, so as we expected, this PR did not cause it.

**Comment by nmarrs** [comment]
> For me, it's more intuitive to rotate a group of elements about a common point, perhaps the centroid of the selected bounding boxes could work.

Hm like move the position of the elements around a common point vs modify each element's rotation in place based on the delta of rotation being applied to the overall group selection box? I think I understand where you are coming from - I think this initial implementation is probably "good enough" / doesn't appear totally broken to user etc but yeah we can explore improving this for sure 👍 

Implementing this would be a bit more involved with actually modifying the positioning of each element vs just the element's rotation property

**Comment by drew08t** [comment]
> > The rotation works as expected for me, but I'm seeing some strange things happening to styling on multiple panels. Notice the background colors: image Not sure if it's a regression from this PR (doubt it) or something else is happening.
> 
> Hm I noticed regression as well, I think it is pretty recent regression that we probably will want to get to the bottom of sooner than later
> 
> > 1. In general this is a good step forward, but I think we need to revisit group rotation in the near future. For example, should we have all elements rotate about a single point? Obviously the handling of anchors needs to be improved as well in a follow-up.
> 
> How else might a group of elements rotate? 🤔 I think moveable has support for changing the relative "center" of rotation but not sure what use cases / value this would add in this case
> 
> Re: connection anchor rendering issue we have [this issue to follow up on this](https://github.com/grafana/grafana/issues/84957)
> 
> > 2. Why are there two rotation control handles?
> 
> See [this issue](https://github.com/grafana/grafana/issues/84959) and [the PR fix](https://github.com/grafana/grafana/pull/85690) that addressed it :)



> > For me, it's more intuitive to rotate a group of elements about a common point, perhaps the centroid of the selected bounding boxes could work.
> 
> Hm like move the position of the elements around a common point vs modify each element's rotation in place based on the delta of rotation being applied to the overall group selection box? I think I understand where you are coming from - I think this initial implementation is probably "good enough" / doesn't appear totally broken to user etc but yeah we can explore improving this for sure 👍
> 
> Implementing this would be a bit more involved with actually modifying the positioning of each element vs just the element's rotation property

Yep, we would need to apply some transformations, shouldn't be too horrible, but definitely worth exploring and following up on later.

**Pull request merged**